### ISSUE TITLE
unit test "testLuaDestructors" failure on many architecures other than amd64/i386.

### ISSUE THREAD

**Issue by coldtobi** [minetest/minetest#12778]
##### Minetest version

```
(sid_armhf-dchroot)tobi@amdahl:~/minetest_sid_armhf/minetest-5.6.0+dfsg+~1.9.0mt7+dfsg/obj-arm-linux-gnueabihf$ ../bin/minetest --version
Minetest 5.6.0 (Linux)
Using Irrlicht 1.9.0mt7
Using LuaJIT 2.1.0-beta3
BUILD_TYPE=RelWithDebInfo
RUN_IN_PLACE=0
USE_CURL=1
USE_GETTEXT=1
USE_SOUND=1
STATIC_SHAREDIR="/usr/share/games/minetest"
STATIC_LOCALEDIR="/usr/share/locale"
```

##### OS / Hardware

Operating system:
```cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux bookworm/sid"
NAME="Debian GNU/Linux"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

CPU:
```
(sid_armhf-dchroot)tobi@amdahl:~/minetest_sid_armhf/minetest-5.6.0+dfsg+~1.9.0mt7+dfsg/obj-arm-linux-gnueabihf$ cat /proc/cpuinfo | head
processor       : 0
model name      : ARMv8 Processor rev 1 (v8l)
BogoMIPS        : 100.00
Features        : half thumb fastmult vfp edsp neon vfpv3 tls vfpv4 idiva idivt lpae evtstrm
CPU implementer : 0x50
CPU architecture: 8
CPU variant     : 0x0
CPU part        : 0x000
CPU revision    : 1
```

###### Summary

Recently, when packaging the new minetest version for Debian, I've enabled unit tests to be run
during build. This uncovered a test failure in TestLua::testLuaDestructors on many architectures Debian builds for. [1]
The build logs for the upload in Debian can be found [here](https://buildd.debian.org/status/logs.php?pkg=minetest&ver=5.6.0%2Bdfsg%2B%7E1.9.0mt7%2Bdfsg-1&suite=sid), to retrieve the build logs click the red "Maybe-Failed" links.

I've debugged this issue on a Debian armhf porter box, and found an issue in TestLua::testLuaDestructors.
There, lua_cpcall() is called with a lamda. The lambda creates a local DestructorDetector object, and then luaL_error() is called.

However as the luaL_error() never returns [2], so the local object never leaves scope and will never be destroyed.
The patch puts the test object in a scope, so that the compiler is allowed to clean up the object.

(The lua documentation tells that luaL_error should be always used like "return LuaL_error(…)", however, this seems not to make
an difference here)

[1] This unit tests interestingly fails only on most of the architectures, but not on the amd64/i386 release architectures.
They also don't fail on some ports, like m68k and sh4). Possibly explanation is in [2]
[2] documentation says that it will do a longjmp(), at least if lua believes that it is in C code. (which is seems, as luajit is included as "extern C"
With that, from https://cplusplus.com/reference/csetjmp/longjmp/ 
"In C++, the implementation *may* perform stack unwinding that destroys objects with automatic duration. If this invokes any non-trivial destructors, it causes *undefined behavior*." As DestructorDetector has a non-trivial dtor, we might be in undefined behaviour land.

That luaL_error is also used elsewhere, I did not check if it is used in a way to avoiding undefined behavior, so this could hide a valid finding, so I'm filing that
as bug report instead of providing a patch that just put the destructor test into a scope.

I'm still trying to completly understand the issue, for example if there is an way to tell luaJIT to use C++. (the header lua.hpp in the Debian luaJIT package just wraps the C-headers in in extern "C" statements as well.

##### Steps to reproduce

(Compile on Debian armhf, and run minetest --run-unittests)

**Comment by sfan5** [comment]
> However as the luaL_error() never returns [2], so the local object never leaves scope and will never be destroyed.
The patch puts the test object in a scope, so that the compiler is allowed to clean up the object.

If you patch the test by moving the definition all you do is break the test and create a false-positive.
The test replicates a situation that exists countless times inside Minetest and cannot be fixed in any other way.

> I'm still trying to completly understand the issue, for example if there is an way to tell luaJIT to use C++. (the header lua.hpp in the Debian luaJIT package just wraps the C-headers in in extern "C" statements as well.

For LuaJIT this table explains whether it's supported: https://luajit.org/extensions.html#exceptions<b></b>¹
C++ destructor interoperability does not work at all with PUC Lua.
So in conclusion the test currently on succeeds on x64 (and i386 apparently) when using LuaJIT.

¹: This is assuming you have a recent LuaJIT from git. These issues are even worse with 2.0 or 2.1.0-beta4.

<hr/>

Effects you can expect if the test fails: Minetest continues to work mostly normally, but memory leaks can happen and error handling might not work correctly in edge cases.
For this reason building with system Lua will become impossible with 5.7.0 as Minetest will ship a minimally modified version that fixes these interoperability issues.

**Comment by coldtobi** [comment]
It was clear to me that the patch this will only hide the issue, but I had forgotten to remove the sentence. Sorry for that confusion.

Thanks for the explaining the effects. This makes me more easy to disable the test for now. (otherwise minetest will not be part of the next Debian release.)

> So in conclusion the test currently only succeeds on x64 (and i386 apparently) when using LuaJIT.
Did you mean "it succedes ONLY BECAUSE of using LuaJIT, or DESPITE of using LuaJIT?" (I'm not a native speaker and I can read it both ways.)  IOW, do you think would falling back to non-JIT make it worse or better?

Said that, I have a mixture of failures on architectures using JIT and not JIT. e.g arm64 hppa and mipsel fails, but it is not using the luaJIT; e.g mipsel, armel/armhf failing using the JIT. (the lists  are not complete)
A few builds succeeds:  amd64/i386 are ok with JIT, and the m68k and sh4ports are ok with non-JIT…

> For this reason building with system Lua will become impossible with 5.7.0 as Minetest will ship a minimally modified version that fixes these interoperability issues 

Also thanks for that information. From a distribution perspective, this is unfortunate news, as in Debian we try hard to use avoid [embedded code copies](https://wiki.debian.org/EmbeddedCopies). But I understand your side, and as lua5.1 is not longer maintained upstream, and upgrading to later lua won't work e.g due to JIT only available for 5.1, there's no way for you to otherwise fix such issues… However, let me formulate a wish: Please check your embedded code copy that it fixes the known vulnerabilities; [Debian's security tracker lists a few](https://security-tracker.debian.org/tracker/source-package/lua5.1); I've only briefly checked, but I think [CVE-2014-5461](https://security-tracker.debian.org/tracker/CVE-2014-5461) and possibly [CVE-2021-43519](https://security-tracker.debian.org/tracker/CVE-2021-43519) seems to be not adresses in your copy. 

Cheers and thanks again for the information!
tobi

**Comment by sfan5** [comment]
> Did you mean "it succedes ONLY BECAUSE of using LuaJIT, or DESPITE of using LuaJIT?" (I'm not a native speaker and I can read it both ways.) IOW, do you think would falling back to non-JIT make it worse or better?

It succeeds *thanks to* LuaJIT. In your (Debian's) situation building LuaJIT is the only way the test could potentially suceed.

> A few builds succeeds: amd64/i386 are ok with JIT, and the m68k and sh4ports are ok with non-JIT…

I briefly looked at that and it seems the builders for the last two arches just didn't run any tests...

> However, let me formulate a wish: Please check your embedded code copy that it fixes the known vulnerabilities; [Debian's security tracker lists a few](https://security-tracker.debian.org/tracker/source-package/lua5.1); I've only briefly checked, but I think [CVE-2014-5461](https://security-tracker.debian.org/tracker/CVE-2014-5461) and possibly [CVE-2021-43519](https://security-tracker.debian.org/tracker/CVE-2021-43519) seems to be not adresses in your copy.

Thanks for the advice. I'll make a note to look at that later.

**Comment by coldtobi** [comment]
> It succeeds _thanks to_ LuaJIT. In your (Debian's) situation building LuaJIT is the only way the test could potentially suceed.

Thanks, that's helping. 
I've configured the Debian package to use luaJIT where it is available (amd64 arm64 armel armhf i386 mips64el mipsel s390x hurd-i386), and disabled the testLuaDestructors. So minetest 5.6.0 should migrate to testing in a few days.

I've seen that there is a 5.6.1, but did not find time to take a closer look.

(If you want, this issue can be closed.)

**Comment by tuxayo** [comment]
Does anyone know if the following quote means that on a given linux distribution release, one can check by the tests if the package need to be built with the embedded Lua?

https://github.com/minetest/minetest/issues/12778#issuecomment-1250255332

> Effects you can expect if the test fails: Minetest continues to work mostly normally, but memory leaks can happen and error handling might not work correctly in edge cases.

**Comment by sfan5** [comment]
Basically yes.
Building with system Lua is no longer an option however so you should always receive a working build regardless of your settings.

**Comment by okias** [comment]
I think for Alpine, we need to workaround this test too. Maybe it could get skipped on non-luajit?

**Comment by okias** [comment]
Make sense, thanks for the clarification.

So far no harm, since Alpine didn't have tests in the place, so it was most likely failing even before. I'll revert the patch, when I figure out, how to fix it.
